### PR TITLE
PyO3 0.22

### PR DESCRIPTION
**Pull request by diliop**
Upgrade from PyO3 `0.21` to `0.22`

Closes https://github.com/huggingface/tokenizers/issues/1639

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Nice! 🔥 
Just approved the workflows, let's go 🚀 
Feel free to ping me, tho I think you are up to a great start!

**Comment by yookoala** [comment]
Sadly ran into numpy compilation issue on Windows (32bit).

Details:
https://github.com/PyO3/rust-numpy/issues/448

**Comment by yookoala** [comment]
The fix has merged to rust-numpy. Now we need to wait for a new release that includes the merge.

**Comment by yookoala** [comment]
@diliop: rust-numpy [v0.22.1](https://github.com/PyO3/rust-numpy/releases/tag/v0.22.1) has been released and it should have the Win32 compilation issue fixed. Please update to use it and try again.

**Comment by diliop** [comment]
There is no need to change anything in code. Just re-running checks on this PR is sufficient since `numpy` is already set to `0.22` which includes `0.22.1`. Looking a the checks above, you can see the `0.22.1` download [here](https://github.com/huggingface/tokenizers/actions/runs/11578853026/job/32278732238?pr=1665#step:6:47). That said, everything looks green :)

**Comment by yookoala** [comment]
@diliop: I think you probably should add "Closes #1639" to the PR's description so it will automatically close it when merged.

**Comment by Rishang** [comment]
+1

File: bindings/python/src/models.rs [hunk 1]
```diff
@@ -181,7 +181,7 @@ impl PyModel {
     ///
     /// Returns:
     ///     :obj:`List[str]`: The list of saved files
-    #[pyo3(text_signature = "(self, folder, prefix)")]
+    #[pyo3(signature = (folder, prefix=None, name=None), text_signature = "(self, folder, prefix, name)")]
```
Comment (ArthurZucker) [inline bindings/python/src/models.rs:184]: since the name has been deprecated you can just remove it here instead!

**Review by ArthurZucker — APPROVED** [review]
Thanks a TON! Let's remove the name and good to go

File: bindings/python/src/models.rs (hunk 1 above)
Comment (diliop) [reply bindings/python/src/models.rs]: Removed `name` from `text_signature` and updated Python stubs

**Comment by ArthurZucker** [comment]
🎉 thanks for the PR 🤗

File: bindings/python/src/tokenizer.rs [hunk 2]
```diff
@@ -1030,7 +1030,7 @@ impl PyTokenizer {
     fn encode_batch(
         &self,
         py: Python<'_>,
-        input: Vec<&PyAny>,
+        input: Bound<'_, PyList>,
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply bindings/python/src/tokenizer.rs:1033]: @diliop we need this to be probably PySequence, but I am not sure about the fix

[the remark this answers is not in the document]
Comment (diliop) [reply bindings/python/src/tokenizer.rs:1033]: Hey, I just saw this :( Yeah I "assumed" that `Vec` was only accepting list from Python hence `PyList` but if you need tuples then `PySequence` is indeed the way to go. `Py*` will give you the benefit of the Python type check at almost zero cost as you mentioned in your PR so they should be preferred where possible. That said, there should be tests covering this so I can pick this up over the weekend and make sure anything else I changed is also covered.

[the remark this answers is not in the document]
Comment (ArthurZucker) [reply bindings/python/src/tokenizer.rs:1033]: Cool yeah was in a rush to fix this, forgot about the tests, super nice if you want to add them 🤗

[the remark this answers is not in the document]
Comment (diliop) [reply bindings/python/src/tokenizer.rs:1033]: #1679 should be the fix. Still planning to add more tests but this my hope is would be enough to restore the previous functionality.

**Pull request merged**